synapse/rest/client/profile.py

```diff
@@ -227,19 +230,106 @@ async def on_GET(
         user = UserID.from_string(user_id)
         await self.profile_handler.check_profile_query_allowed(user, requester_user)
 
-        displayname = await self.profile_handler.get_displayname(user)
-        avatar_url = await self.profile_handler.get_avatar_url(user)
+        return 200, await self.profile_handler.get_profile(user_id)
 
-        ret = {}
-        if displayname is not None:
-            ret["displayname"] = displayname
-        if avatar_url is not None:
-            ret["avatar_url"] = avatar_url
 
-        return 200, ret
+class UnstableProfileRestServlet(RestServlet):
+    PATTERNS = [
+        re.compile(
+            r"^/_matrix/client/unstable/uk\.tcpip\.msc4133/profile/(?P<user_id>[^/]*)/(?P<field_name>[^/]*)"
+        )
+    ]
+    CATEGORY = "Event sending requests"
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__()
+        self.hs = hs
+        self.profile_handler = hs.get_profile_handler()
+        self.auth = hs.get_auth()
+
+    async def on_GET(
+        self, request: SynapseRequest, user_id: str, field_name: str
+    ) -> Tuple[int, JsonDict]:
+        requester_user = None
+
+        if self.hs.config.server.require_auth_for_profile_requests:
+            requester = await self.auth.get_user_by_req(request)
+            requester_user = requester.user
+
+        if not UserID.is_valid(user_id):
+            raise SynapseError(
+                HTTPStatus.BAD_REQUEST, "Invalid user id", Codes.INVALID_PARAM
+            )
+
+        user = UserID.from_string(user_id)
+        await self.profile_handler.check_profile_query_allowed(user, requester_user)
+
+        if field_name == ProfileFields.DISPLAYNAME:
+            field_value = await self.profile_handler.get_displayname(user)
+        elif field_name == ProfileFields.AVATAR_URL:
+            field_value = await self.profile_handler.get_avatar_url(user)
+        else:
+            field_value = await self.profile_handler.get_profile_field(user, field_name)
+
+        return 200, {field_name: field_value}
+
+    async def on_PUT(
+        self, request: SynapseRequest, user_id: str, field_name: str
+    ) -> Tuple[int, JsonDict]:
+        requester = await self.auth.get_user_by_req(request)
+        user = UserID.from_string(user_id)
+        is_admin = await self.auth.is_server_admin(requester)
+
+        if len(field_name) > MAX_CUSTOM_FIELD_LEN:
```

Comment:
In Python, `len(str)` for utf-8 encoded strings will count the number of unicode codepoints, not bytes. The MSC specifies that field names should be 255 bytes in length.

You can check the byte length by instead first converting to a byte array:

```suggestion
        if len(field_name.encode("utf-8")) > MAX_CUSTOM_FIELD_LEN:
```

likewise below.

Reply:
I think this check doesn't fully match the MSC since that talks about the fully encoded profile JSON dict.

Reply:
We just [had a long discussion](https://github.com/matrix-org/matrix-spec-proposals/pull/4133#discussion_r1698225294) on size limits, and concluded that 64KiB be the limit for profiles when communicating to other homeservers over federation. Homeservers can then choose to enforce that limit on the client side or not. Currently there's no defined method of negotiation, so unfortunately clients will have to try and handle `M_TOO_LARGE` responses for now.

The 255 byte limit for field names remains, but the full profile limit is missing.

For Synapse, I suggest we keep to a 64KiB limit for the whole profile for now. We could compute it entirely in the DB, which may be faster than pulling the fields out of the DB, and then computing their length in Python. Something like (untested):

PostgreSQL:

```sql
SELECT LENGTH(to_jsonb(jsonb_object_agg(name, value))::text) AS json_byte_length 
FROM profile_fields 
WHERE
  user_id = ?;
```

SQLite:

```sql
WITH json_data AS (
    SELECT json_group_object(name, value) AS json_blob
    FROM profile_fields
    WHERE user_id = ?
)
SELECT LENGTH(json_blob) AS json_byte_length
FROM json_data;
```

(SQLite doesn't support `LENGTH` function on JSON.)